contracts/gol.cairo

```diff
@@ -0,0 +1,450 @@
+        generation=generation,
+        game_state=INFINITE_GAME_GENESIS,
+        user_id=caller
+    )
+
+    ERC20.initializer(token_name, token_symbol, token_decimals)
+    Proxy.initializer(proxy_admin)
+    return ()
+end
+
+# Create a new game (creator mode)
+@external
+func create{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game_state : felt
+    ):
+    alloc_locals
+    let (caller) = ensure_user()
+    verify_new_game(game_state)
+    pay(user=caller, credit_requirement=CREATE_CREDIT_REQUIREMENT)
+    save_game(
+        generation=0,
+        game_state=game_state,
+        user_id=caller
+    )    
+    return ()
+end
+
+# Progresses the game by one generation
+@external
+func evolve{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game_id : felt
+    ):
+    alloc_locals
+    # evolve by one generation
+    const generations = 1
+
+    let (caller) = ensure_user()
+    evolve_and_save(
+        game_id=game_id,
+        user=caller,
+        generations=generations
+    ) 
+    reward_user(user=caller)
+
+    return ()
+end
+
+
+# Give life to a specific cell (infinite game).
+@external
+func give_life_to_cell{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        cell_index : felt
+    ):
+    # This does not trigger an evolution. Multiple give_life
+    # operations may be called, building up a shape before
+    # a turn triggers evolution.
+    alloc_locals
+    let (caller) = ensure_user()
+    let (generation, current_game_state) = verify_and_get_for_cell_activation(
+        cell_index=cell_index,
+        user=caller
+    )
+    pay(user=caller, credit_requirement=GIVE_LIFE_CREDIT_REQUIREMENT)
+    activate_cell(
+        generation=generation,
+        caller=caller,
+        cell_index=cell_index,
+        current_state=current_game_state
+    )
+
+    return ()
+end
+
+##########
+
+# Returns a felt of encoded cells for the specified game's generation.
+@view
+func view_game{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game_id : felt,
+        generation : felt
+    ) -> (
+        generation_owner : felt,
+        game_state : felt
+    ):
+
+    let (game_info) = stored_game.read(game_id, generation)
+    return (game_info[0], game_info[1])
+end
+
+# Returns the current generation of a given game.
+@view
+func get_current_generation{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game_id : felt
+    ) -> (
+        generation : felt
+    ):
+    let (generation) = current_generation.read(game_id)
+    return (generation)
+end
+
+##########
+
+func pay{
```

Comment:
Please move all helper function to a separate file, leaving only interface in here.